flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/BulkFormatMapping.java

```diff
@@ -0,0 +1,133 @@
+
+import org.apache.flink.connector.file.src.FileSourceSplit;
+import org.apache.flink.connector.file.src.reader.BulkFormat;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.store.file.KeyValue;
+import org.apache.flink.table.store.file.predicate.Predicate;
+import org.apache.flink.table.store.file.schema.DataField;
+import org.apache.flink.table.store.file.schema.KeyValueFieldsExtractor;
+import org.apache.flink.table.store.file.schema.RowDataType;
+import org.apache.flink.table.store.file.schema.SchemaEvolutionUtil;
+import org.apache.flink.table.store.file.schema.TableSchema;
+import org.apache.flink.table.store.format.FileFormat;
+import org.apache.flink.table.store.utils.Projection;
+import org.apache.flink.table.types.logical.RowType;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+/** Class with index mapping and bulk format. */
+public class BulkFormatMapping {
+    @Nullable private final int[] indexMapping;
+    private final BulkFormat<RowData, FileSourceSplit> bulkFormat;
+
+    public BulkFormatMapping(int[] indexMapping, BulkFormat<RowData, FileSourceSplit> bulkFormat) {
+        this.indexMapping = indexMapping;
+        this.bulkFormat = bulkFormat;
+    }
+
+    @Nullable
+    public int[] getIndexMapping() {
+        return indexMapping;
+    }
+
+    public BulkFormat<RowData, FileSourceSplit> getReaderFactory() {
+        return bulkFormat;
+    }
+
+    public static BulkFormatMappingBuilder newBuilder(
+            FileFormat fileFormat,
+            KeyValueFieldsExtractor extractor,
+            int[][] keyProjection,
+            int[][] valueProjection,
+            int[][] projection,
+            @Nullable List<Predicate> filters) {
+        return new BulkFormatMappingBuilder(
+                fileFormat, extractor, keyProjection, valueProjection, projection, filters);
+    }
+
+    /** Builder to build {@link BulkFormatMapping}. */
+    public static class BulkFormatMappingBuilder {
+        private final FileFormat fileFormat;
+        private final KeyValueFieldsExtractor extractor;
+        private final int[][] keyProjection;
+        private final int[][] valueProjection;
+        private final int[][] projection;
+        @Nullable private final List<Predicate> filters;
+
+        private BulkFormatMappingBuilder(
+                FileFormat fileFormat,
+                KeyValueFieldsExtractor extractor,
+                int[][] keyProjection,
+                int[][] valueProjection,
+                int[][] projection,
+                @Nullable List<Predicate> filters) {
+            this.fileFormat = fileFormat;
+            this.extractor = extractor;
+            this.keyProjection = keyProjection;
+            this.valueProjection = valueProjection;
+            this.projection = projection;
+            this.filters = filters;
+        }
+
+        public BulkFormatMapping build(TableSchema tableSchema, TableSchema dataSchema) {
```

Comment:
Can we simplify this process? For example, we can build new special KeyValue fieldIds based on all fields of KeyValue, so that this process can be similar to Append only, and should be simplified.
What do you think?